### DISCUSSION TITLE
Is there any relationship between compaction and search performance?

### DISCUSSION THREAD

**Discussion by hjlee9182** [milvus-io/milvus#30653] [Q&A and General discussion]
I have 2 collections.
A collection is loaded and searched. B collection is not loaded and upserted.
Each collection have almost 2 Million entity and 512 emb.

While I was upserting to B collection, the search performance of A collection fell. When I upserted in a lot at once, the performance went up to 200-300ms, so I put 1 second on the sleep when each 100 entities upserted.
And when I checked the metric, the search performance also fell whenever a compaction occurred in the data coordinator.

1. What is the connection between these two?
2. How can I lower the compaction latency? Do I just need to increase the data node?
![image](https://github.com/milvus-io/milvus/assets/46337218/80d63360-1b66-4853-9287-c6cc00f63bb1)
![image](https://github.com/milvus-io/milvus/assets/46337218/d2c18e65-c39f-4301-9d4b-a51dcf6371c1)
![image](https://github.com/milvus-io/milvus/assets/46337218/c590db7a-da86-4959-86cc-c10c045b91d1)

**Comment by yhmo** [comment]
1. double-check the "Segment Loaded Num" and "Collection Loaded Num" in the "Query Node" section
the "Collection Loaded Num" should be 1
the "Segment Loaded Num" should be a const value
![Screenshot from 2024-02-19 15-18-10](https://github.com/milvus-io/milvus/assets/2282099/85e88c1d-ac08-4d8c-9c2d-b6ea85c3ce7d)


2. double-check the "Search Vector Count Rate" in the "Proxy" section, it should be almost a const value
![Screenshot from 2024-02-19 15-18-56](https://github.com/milvus-io/milvus/assets/2282099/fb2dda78-70d9-4a2e-9ae3-026aa0a96903)

**Comment by hjlee9182** [comment]
Use Milvus 2.3.4-gpu

Query Node  - Segment Loaded Num, Collection Loaded Num

![image](https://github.com/milvus-io/milvus/assets/46337218/3c21b900-4ec8-4dc8-92e3-82e40401d579)


Proxy - Search Vecotr Count Rate
I read it using locust

![image](https://github.com/milvus-io/milvus/assets/46337218/a4d0a2ce-79af-4a47-bf37-d810792cac85)

**Reply by yhmo** [reply]
Looks like the "Search Vector Count Rate" is affected by upsert at 14:18, 14:28 and 14:54

Check the "Mutation Latency" and "Insert Vector Count Rate" in "Proxy" section.
![Screenshot from 2024-02-19 16-06-45](https://github.com/milvus-io/milvus/assets/2282099/9ab49268-5cb3-42a5-b863-225e67aebe88)

**Reply by hjlee9182** [reply]
here!
![image](https://github.com/milvus-io/milvus/assets/46337218/170141b3-fadf-4cc0-a71a-532c00a0fffd)

But In my metric, upsert is not show in insert vector count Rate only insert is show.

![image](https://github.com/milvus-io/milvus/assets/46337218/1b352986-6ac4-4b20-b612-fb23b8c9180a)

**Reply by yhmo** [reply]
Yes, upsert is not counted in the "Insert Vector Count Rate".
Search latency is affected by upsert requests in proxy node. There are three peaks of "Stored Rows Rate" at 14:18, 14:28 and 14:54. Proxy node needs to handle more requests(search + upsert) at these points.

You have at least 3 proxy nodes, check your client code, use different proxy nodes for different requests. For example, all search requests are sent to proxy_1 or proxy_2, all upsert requests are sent to proxy_3.

**Reply by hjlee9182** [reply]
I tested  use different proxy nodes for different requests. But that's not effect. Same situation again occured.
I used the code below.
``` python

Host  = '{ proxy pod IP }'
port = 19530

connections.connect(host=Host, port=Port)

```

![image](https://github.com/milvus-io/milvus/assets/46337218/29d94007-84c5-4cf7-bfb6-d7bcb625d7b0)

**Comment by xiaofan-luan** [comment]
what milvus version are u using && what consistency are you using?

did you check the cpu usage or all your node on metrics?

**Reply by hjlee9182** [reply]
2.3.4-gpu and consistency is bounded.
cpu and all node is working!

**Reply by xiaofan-luan** [reply]
what is the cpu usage?

**Reply by xiaofan-luan** [reply]
I guess this probably due to the GPU queue policy.

**Reply by hjlee9182** [reply]
data node
![image](https://github.com/milvus-io/milvus/assets/46337218/90714b57-432c-43f0-8c53-cba8480ce3d2)

index node
![image](https://github.com/milvus-io/milvus/assets/46337218/03141805-d9ba-4245-850d-a29c8dc8e77c)

proxy

![image](https://github.com/milvus-io/milvus/assets/46337218/e4106d45-952d-4dd7-a334-54eed9750831)

query node
![image](https://github.com/milvus-io/milvus/assets/46337218/7a47c9cb-466b-43de-84f2-38124313263d)

**Reply by hjlee9182** [reply]
But this phenomenon also happened in version 2.3.4milvus ( no gpu).

**Reply by hjlee9182** [reply]
and test in 2.3.9, also occured.

**Reply by barzan-hayati** [reply]
@hjlee9182  how you generated these charts?

**Reply by yhmo** [reply]
@barzan-hayati https://milvus.io/docs/monitor.md

[the remark this answers is not in the document]
**Reply by xiaofan-luan** [reply]
index build take GPU which slow the search?
Did you deploy index and querynode on same GPU card?

[the remark this answers is not in the document]
**Reply by hjlee9182** [reply]
diffrent card

**Comment by xiaofan-luan** [comment]
The cpu usgae of both proxy and querynode seems to be high.

How many qps are you running during the test and what's the target QPS/latency?

If you are using GPU I guess cpu should't be taken too much?

Maybe we can have a quick online meeting sync?

**Reply by hjlee9182** [reply]
I tested about 1000 qps,.
I don't have an exact target qps, but I'm testing it to see how well it performs.
Maybe I have GPU, but not using gpu index(ex. GPU_IV_FLAT) so cpu was used too much.

I'm Ok, but what purpose of quick online meeting?

**Reply by hjlee9182** [reply]
Is this not normally situation?

**Reply by xiaofan-luan** [reply]
If you need to any futher help of improving performance, we are open to do so.

From what all you described, there must be some shared resources of insert of seaerch. My guess
1. proxy cpu
2. proxy network bandwith(very likely)
3. GPU card, you said index and search are not sharing resources so this is not likely.

Maybe check network bandwidth. This happens especially when you retrieve fields out
